### PR TITLE
C#: Query to detect hash without salt

### PR DESCRIPTION
**Pull request by luchua-bc**
In cryptography, a salt is some random data used as an additional input to a one-way function that hashes a password or pass-phrase. It makes dictionary attacks more difficult.

Without a salt, it is much easier for attackers to pre-compute the hash value using dictionary attack techniques such as rainbow tables to crack passwords.

This type of vulnerabilities is categorized as CWE-759 "Use of a One-Way Hash without a Salt". This query detects issues of this type.

Please consider to merge the PR. Thanks.

### CODE REVIEW

File: csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
+// semmle-extractor-options: /r:System.Security.Cryptography.Primitives.dll /r:System.Security.Cryptography.Csp.dll /r:System.Security.Cryptography.Algorithms.dll
```
Comment (hvitved) [inline csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.cs:1]: Please move this test file, along with the `.expected` file to a new `Security Features/CWE-759` folder under https://github.com/github/codeql/tree/main/csharp/ql/test/experimental. You will also need to add a `.qlref` file.

File: csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.ql [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  }
+}
+
+/** The method call `ComputeHash()` declared in `System.Security.Cryptography.SHA...`. */
+class HashDataMethodCall extends MethodCall {
+  HashDataMethodCall() {
+    this.getQualifier().getType() instanceof HashAlgorithmProvider and
+    this.getTarget().hasName("HashData")
+  }
+}
```
Comment (hvitved) [inline csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.ql:39]: I would restructure this a bit:
```ql
class HashMethod extends Method {
  HashMethod() {
    this.getDeclaringType() instanceof SHA and
    this.hasName("ComputeHash")
    or
    this.getDeclaringType() instanceof HashAlgorithmProvider and
    this.hasName("HashData")
  }
}

...

  override predicate isSink(DataFlow::Node sink) {
    exists(MethodCall mc |
      sink.asExpr() = mc.getArgument(0) and
      mc.getTarget() instanceof HashMethod
    )
  }
```

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks very interesting; a few comments.

File: csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.ql (hunk 2 above)
Comment (luchua-bc) [reply csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.ql]: Thanks @hvitved for reviewing this PR. I've made requested changes.

File: csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.cs (hunk 1 above)
Comment (luchua-bc) [reply csharp/ql/src/experimental/Security Features/CWE-759/HashWithoutSalt.cs]: Moved the test files and added a <code>.qlref</code> file

**Comment by hvitved** [comment]
Test execution fails:
```
> codeql test run csharp/ql/test/experimental/Security\ Features/CWE-759/
--- expected
+++ actual
@@ -1,13 +1,7 @@
 edges
 | HashWithoutSalt.cs:17:70:17:77 | access to parameter password : String | HashWithoutSalt.cs:19:49:19:56 | access to local variable passBuff |
-| HashWithoutSalt.cs:37:28:37:72 | call to method GetBytes : Byte[] | HashWithoutSalt.cs:38:51:38:59 | access to local variable passBytes |
-| HashWithoutSalt.cs:37:64:37:71 | access to parameter password : String | HashWithoutSalt.cs:37:28:37:72 | call to method GetBytes : Byte[] |
 nodes
 | HashWithoutSalt.cs:17:70:17:77 | access to parameter password : String | semmle.label | access to parameter password : String |
 | HashWithoutSalt.cs:19:49:19:56 | access to local variable passBuff | semmle.label | access to local variable passBuff |
-| HashWithoutSalt.cs:37:28:37:72 | call to method GetBytes : Byte[] | semmle.label | call to method GetBytes : Byte[] |
-| HashWithoutSalt.cs:37:64:37:71 | access to parameter password : String | semmle.label | access to parameter password : String |
-| HashWithoutSalt.cs:38:51:38:59 | access to local variable passBytes | semmle.label | access to local variable passBytes |
 #select
 | HashWithoutSalt.cs:19:49:19:56 | access to local variable passBuff | HashWithoutSalt.cs:17:70:17:77 | access to parameter password : String | HashWithoutSalt.cs:19:49:19:56 | access to local variable passBuff | $@ is hashed without a salt. | HashWithoutSalt.cs:17:70:17:77 | access to parameter password | The password |
-| HashWithoutSalt.cs:38:51:38:59 | access to local variable passBytes | HashWithoutSalt.cs:37:64:37:71 | access to parameter password : String | HashWithoutSalt.cs:38:51:38:59 | access to local variable passBytes | $@ is hashed without a salt. | HashWithoutSalt.cs:37:64:37:71 | access to parameter password | The password |
```

Also, the `.qhelp` file is now referencing a file that does not exist. You can check the `.qhelp` file with
```
codeql generate query-help --format=markdown csharp/ql/src/experimental/Security\ Features/CWE-759/HashWithoutSalt.qhelp
```

**Comment by luchua-bc** [comment]
@hvitved Sorry about that. I had an issue with my local environment in the last several days, which was just fixed with the latest CLI and QL libraries. Now I can run the test and correct the query. Please review again.

**Review by hvitved — DISMISSED** [review]

**Review by hvitved — APPROVED** [review]
Sorry for the long wait on this one.

**Comment by luchua-bc** [comment]
Thanks @hvitved a lot for the update.

**Review by hvitved — APPROVED** [review]

**Pull request merged**